File: src/main/java/com/baulsupp/oksocial/commands/ShellCommand.java
package com.baulsupp.oksocial.commands;

import com.baulsupp.oksocial.completion.ArgumentCompleter;
import com.baulsupp.oksocial.completion.UrlCompleter;
import java.util.List;
import java.util.Optional;
import okhttp3.OkHttpClient;
import okhttp3.Request;

import static java.util.Optional.empty;

public interface ShellCommand {
  String name();

  List<Request> buildRequests(OkHttpClient client, Request.Builder requestBuilder,
      List<String> arguments) throws Exception;

  default Optional<String> authenticator() {
    return empty();
  }

  default boolean handlesRequests() {
    return false;
  }

  default Optional<ArgumentCompleter> completer() {
    return empty();
  }
}

File: src/main/java/com/baulsupp/oksocial/commands/ShellCommand.kt
package com.baulsupp.oksocial.commands

import com.baulsupp.oksocial.completion.ArgumentCompleter
import java.util.Optional
import okhttp3.OkHttpClient
import okhttp3.Request

import java.util.Optional.empty

interface ShellCommand {
    fun name(): String

    @Throws(Exception::class)
    fun buildRequests(client: OkHttpClient, requestBuilder: Request.Builder,
                      arguments: List<String>): List<Request>

    open fun authenticator(): Optional<String> {
        return empty()
    }

    open fun handlesRequests(): Boolean {
        return false
    }

    fun completer(): Optional<ArgumentCompleter> {
        return empty()
    }
}
